cmd/init.go

```diff
@@ -54,126 +72,144 @@ to quickly create a Cobra application.`,
-		log.Println("clone and detokenization of metaphor-template repository complete")
-		progressPrinter.IncrementTracker("step-metaphor", 1)
+		go func() {
+			//! tracker 0
+			log.Println("installing kubefirst dependencies")
+			progressPrinter.IncrementTracker("step-download", 1)
+			err := downloadManager.DownloadTools(config)
+			if err != nil {
+				//return err
+			}
+			log.Println("dependency installation complete")
+			progressPrinter.IncrementTracker("step-download", 1)
+
+			wg.Done()
+		}()
+
+		go func() {
+			//! tracker 1
+			log.Println("getting aws account information")
+			aws.GetAccountInfo()
+			log.Printf("aws account id: %s\naws user arn: %s", viper.GetString("aws.accountid"), viper.GetString("aws.userarn"))
+			progressPrinter.IncrementTracker("step-account", 1)
+
+			//! tracker 2
+			// hosted zone id
+			// So we don't have to keep looking it up from the domain name to use it
+			hostedZoneId := aws.GetDNSInfo(awsFlags.HostedZoneName)
+			// viper values set in above function
+			log.Println("hostedZoneId:", hostedZoneId)
+			progressPrinter.IncrementTracker("step-dns", 1)
+
+			//! tracker 3
+			// todo: this doesn't default to testing the dns check
+			skipHostedZoneCheck := viper.GetBool("init.hostedzonecheck.enabled")
+			if !skipHostedZoneCheck {
+				aws.TestHostedZoneLiveness(globalFlags.DryRun, awsFlags.HostedZoneName, hostedZoneId)
+			} else {
+				log.Println("skipping hosted zone check")
+			}
+			progressPrinter.IncrementTracker("step-live", 1)
+			wg.Done()
+		}()
+
+		go func() {
+			//! tracker 4
+			//* should we consider going down to a single bucket
+			//* for state and artifacts on open source?
+			//* hitting a bucket limit on an install might deter someone
+			log.Println("creating buckets for state and artifacts")
+			aws.BucketRand(globalFlags.DryRun)
+			progressPrinter.IncrementTracker("step-buckets", 1)
+			log.Println("BucketRand() complete")
+
+			wg.Done()
+		}()
+
+		go func() {
+			//! tracker 5
+			log.Println("creating an ssh key pair for your new cloud infrastructure")
+			pkg.CreateSshKeyPair()
+			log.Println("ssh key pair creation complete")
+			progressPrinter.IncrementTracker("step-ssh", 1)
+
+			wg.Done()
+		}()
+
+		go func() {
+			//! tracker 6
+			prepareKubefirstTemplateRepo(config, viper.GetString("gitops.owner"), viper.GetString("gitops.repo"), viper.GetString("gitops.branch"), viper.GetString("template.tag"))
+			log.Println("clone and detokenization of gitops-template repository complete")
+			progressPrinter.IncrementTracker("step-gitops", 1)
+
+			wg.Done()
+		}()
+
+		go func() {
+			//! tracker 7
+			log.Printf("cloning and detokenizing the metaphor-template repository")
+			prepareKubefirstTemplateRepo(config, viper.GetString("gitops.owner"), "metaphor", "", viper.GetString("template.tag"))
+			log.Println("clone and detokenization of metaphor-template repository complete")
+			progressPrinter.IncrementTracker("step-metaphor", 1)
+
+			wg.Done()
+		}()
+
+		wg.Wait()
```

Comment:
Did you checked if `viper.WriteConfig()` and viper is thread safe? As we have several thread updating values on it. Do we need any mutex protection?

Reply:
Maybe this could help for details: https://github.com/spf13/viper/issues/19